<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU18_Infraestructuras_eventos</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 05. Agrupar datos de POIs por distrito

* A partir de los datos puntuales, agregar por distrito censal y contar POIs para hacer mapas de regiones y usar en modelos
* Adicionalmente crear metadatos con descripción y agrupamiento de variables.








<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Packages to use

- {tcltk} para selección interactiva de archivos locales
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos
- {dityr} para tranformar datos
- {janitor} para limpiar datos


In [1]:
library(readr)
library(dplyr)
library(tidyr)
library(janitor)



Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union



Attaching package: 'janitor'


The following objects are masked from 'package:stats':

    chisq.test, fisher.test




### Paths

In [2]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [ ]:
# file_data <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [3]:
iFile <- "CU_18_05_12_pois_distrito.csv"
file_data <- paste0(iPath, iFile)

if(file.exists(file_data)){
    cat("Se leerán datos del archivo: ", file_data)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/CU_18_05_12_pois_distrito.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [4]:
data <- read_csv(file_data)


Rows: 24780 Columns: 7
-- Column specification --------------------------------------------------------
Delimiter: ","
chr (5): grupo, tipo, nombre, CMUN, CDIS
dbl (2): X, Y

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [5]:
glimpse(data)

Rows: 24,780
Columns: 7
$ grupo  <chr> "turismo", "hosteleria", "hosteleria", "hosteleria", "comercio"~
$ tipo   <chr> "hotel", "restaurant", "pub", "pub", "supermarket", "fast_food"~
$ nombre <chr> "NH Ciudad de la Imagen", "Caf<U+00E9> Comercial", "Sidrer<U+00ED>a la Camoch~
$ X      <dbl> -3.788176, -3.702002, -3.701686, -3.696329, -3.706888, -3.60722~
$ Y      <dbl> 40.39844, 40.42873, 40.42703, 40.42760, 40.48035, 40.43337, 40.~
$ CMUN   <chr> "115", "079", "079", "079", "079", "079", "079", "079", "079", ~
$ CDIS   <chr> "01", "01", "01", "01", "08", "20", "01", "01", "01", "01", "01~


Muestra de datos:

In [6]:
slice_head(data, n = 5)

grupo      tipo        nombre                     X         Y        CMUN
1 turismo    hotel       NH Ciudad de la Imagen     -3.788176 40.39844 115 
2 hosteleria restaurant  Caf<U+00E9> Comercial      -3.702002 40.42873 079 
3 hosteleria pub         Sidrer<U+00ED>a la Camocha -3.701686 40.42703 079 
4 hosteleria pub         Gran Cafe Santander        -3.696329 40.42760 079 
5 comercio   supermarket Alcampo                    -3.706888 40.48035 079 
  CDIS
1 01  
2 01  
3 01  
4 01  
5 08

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* POIs

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

#### Data Transform

* Contar número de infraestructuras de cada tipo por distrito
* Extender en columnas para caracterizar distritos

In [7]:
tdata_01 <- data |> 
  count(CMUN, CDIS, tipo) |> 
  pivot_wider(names_from = tipo,
              values_from = n, 
              values_fill = 0) 

In [8]:
glimpse(tdata_01)

Rows: 224
Columns: 60
$ CMUN              <chr> "002", "003", "004", "005", "005", "005", "005", "00~
$ CDIS              <chr> "01", "01", "01", "01", "02", "03", "04", "05", "01"~
$ bakery            <int> 2, 0, 0, 4, 5, 1, 2, 7, 4, 1, 1, 0, 2, 0, 0, 1, 0, 0~
$ bar               <int> 2, 0, 1, 22, 10, 2, 3, 11, 7, 1, 1, 2, 6, 0, 2, 17, ~
$ butcher           <int> 1, 0, 0, 2, 2, 0, 1, 2, 0, 0, 1, 0, 0, 0, 0, 2, 0, 0~
$ cafe              <int> 5, 1, 0, 12, 6, 1, 3, 19, 12, 3, 2, 0, 6, 0, 1, 4, 0~
$ car_dealership    <int> 1, 0, 0, 1, 1, 3, 0, 3, 5, 1, 0, 0, 0, 0, 0, 0, 0, 0~
$ clothes           <int> 1, 0, 0, 13, 1, 0, 1, 28, 0, 0, 0, 0, 0, 0, 1, 1, 0,~
$ convenience       <int> 2, 0, 0, 4, 2, 2, 2, 13, 8, 1, 1, 2, 6, 0, 0, 1, 0, ~
$ fast_food         <int> 1, 0, 0, 8, 1, 2, 5, 11, 10, 7, 4, 0, 6, 0, 2, 5, 0,~
$ food_court        <int> 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0~
$ furniture_shop    <int> 2, 0, 0, 3, 0, 0, 1, 8, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0~
$ gift_shop       

In [9]:
tdata_01 |> slice_head(n = 5)

CMUN CDIS bakery bar butcher cafe car_dealership clothes convenience
1 002  01   2       2  1        5   1               1      2          
2 003  01   0       0  0        1   0               0      0          
3 004  01   0       1  0        0   0               0      0          
4 005  01   4      22  2       12   1              13      4          
5 005  02   5      10  2        6   1               1      2          
  fast_food ... vending_parking chalet shelter motel camp_site general
1 1         ... 0               0      0       0     0         0      
2 0         ... 0               0      0       0     0         0      
3 0         ... 0               0      0       0     0         0      
4 8         ... 0               0      0       0     0         0      
5 1         ... 0               0      0       0     0         0      
  alpine_hut video_shop vending_machine car_sharing
1 0          0          0               0          
2 0          0          0               0          
3 0          0          0               0          
4 0          0          0               0          
5 0          0          0               0

#### Data Extract




* Extraer nombres de columnas


In [10]:
tdata_02 <- data.frame(desc_var = colnames(tdata_01)[3:ncol(tdata_01)])

In [11]:
tdata_02

desc_var         
1  bakery           
2  bar              
3  butcher          
4  cafe             
5  car_dealership   
6  clothes          
7  convenience      
8  fast_food        
9  food_court       
10 furniture_shop   
11 gift_shop        
12 greengrocer      
13 guesthouse       
14 hairdresser      
15 hotel            
16 pub              
17 restaurant       
18 stationery       
19 vending_any      
20 bicycle_rental   
21 hostel           
22 supermarket      
23 beauty_shop      
24 beverages        
25 bicycle_shop     
26 bookshop         
27 chemist          
28 computer_shop    
29 doityourself     
30 florist          
31 jeweller         
32 kiosk            
33 mall             
34 mobile_phone_shop
35 optician         
36 shoe_shop        
37 toy_shop         
38 travel_agent     
39 car_rental       
40 car_wash         
41 outdoor_shop     
42 sports_shop      
43 caravan_site     
44 biergarten       
45 garden_centre    
46 laundry          
47 department_store 
48 newsagent        
49 vending_parking  
50 chalet           
51 shelter          
52 motel            
53 camp_site        
54 general          
55 alpine_hut       
56 video_shop       
57 vending_machine  
58 car_sharing

Data Transform

* Completar metadatos

In [13]:
tdata_02 <- tdata_02 |>
    mutate(nombre_var = colnames(tdata_01)[3:ncol(tdata_01)]) |>
    left_join(data |> count(grupo, tipo), 
                by = c("desc_var" = "tipo")) 

In [14]:
glimpse(tdata_02)

Rows: 58
Columns: 4
$ desc_var   <chr> "bakery", "bar", "butcher", "cafe", "car_dealership", "clot~
$ nombre_var <chr> "bakery", "bar", "butcher", "cafe", "car_dealership", "clot~
$ grupo      <chr> "comercio", "hosteleria", "comercio", "hosteleria", "comerc~
$ n          <int> 627, 2044, 289, 1577, 261, 1435, 1200, 1049, 24, 286, 129, ~


In [15]:
tdata_02

desc_var          nombre_var        grupo      n   
1  bakery            bakery            comercio    627
2  bar               bar               hosteleria 2044
3  butcher           butcher           comercio    289
4  cafe              cafe              hosteleria 1577
5  car_dealership    car_dealership    comercio    261
6  clothes           clothes           comercio   1435
7  convenience       convenience       comercio   1200
8  fast_food         fast_food         hosteleria 1049
9  food_court        food_court        hosteleria   24
10 furniture_shop    furniture_shop    comercio    286
11 gift_shop         gift_shop         comercio    129
12 greengrocer       greengrocer       comercio    548
13 guesthouse        guesthouse        turismo     285
14 hairdresser       hairdresser       comercio   1366
15 hotel             hotel             turismo     340
16 pub               pub               hosteleria 1201
17 restaurant        restaurant        hosteleria 5772
18 stationery        stationery        comercio    190
19 vending_any       vending_any       comercio     55
20 bicycle_rental    bicycle_rental    comercio    235
21 hostel            hostel            turismo      90
22 supermarket       supermarket       comercio   1135
23 beauty_shop       beauty_shop       comercio    358
24 beverages         beverages         comercio     62
25 bicycle_shop      bicycle_shop      comercio    146
26 bookshop          bookshop          comercio    252
27 chemist           chemist           comercio     60
28 computer_shop     computer_shop     comercio    131
29 doityourself      doityourself      comercio    354
30 florist           florist           comercio    179
31 jeweller          jeweller          comercio    200
32 kiosk             kiosk             comercio    255
33 mall              mall              comercio     19
34 mobile_phone_shop mobile_phone_shop comercio    213
35 optician          optician          comercio    339
36 shoe_shop         shoe_shop         comercio    411
37 toy_shop          toy_shop          comercio    116
38 travel_agent      travel_agent      comercio    180
39 car_rental        car_rental        comercio     56
40 car_wash          car_wash          comercio     89
41 outdoor_shop      outdoor_shop      comercio     27
42 sports_shop       sports_shop       comercio    111
43 caravan_site      caravan_site      turismo       4
44 biergarten        biergarten        hosteleria   28
45 garden_centre     garden_centre     comercio     30
46 laundry           laundry           comercio    197
47 department_store  department_store  comercio     32
48 newsagent         newsagent         comercio     62
49 vending_parking   vending_parking   comercio    608
50 chalet            chalet            turismo      19
51 shelter           shelter           turismo      54
52 motel             motel             turismo      11
53 camp_site         camp_site         turismo       4
54 general           general           comercio     12
55 alpine_hut        alpine_hut        turismo       1
56 video_shop        video_shop        comercio      9
57 vending_machine   vending_machine   comercio     12
58 car_sharing       car_sharing       comercio      1

Si no aplica: Estos datos no requieren tareas de este tipo.

## Synthetic Data Generation

Estos datos no requieren tareas de este tipo.

## Fake Data Generation

Estos datos no requieren tareas de este tipo.

## Open Data

Estos datos no requieren tareas de este tipo.

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

<font color='tomato'> Identificamos los datos a guardar</font>

1. Infraestructuras agregadas por distrito

In [16]:
data_to_save_01 <- tdata_01


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [18]:
caso <- "CU_18"
proceso <- '_05'
tarea <- "_13"
archivo <- "_01"
proper <- "_ditrito_pois"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [ ]:
# file_save_01 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_01 <- paste0(oPath, file_save_01)
# write_csv(data_to_save_01, path_out_01)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [19]:
file_save_01 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_01 <- paste0(oPath, file_save_01)
write_csv(data_to_save_01, path_out_01)

cat('File saved as: ')
path_out_01

File saved as: 

[1] "Data/Output/CU_18_05_13_01_ditrito_pois.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [20]:
path_in_01 <- paste0(iPath, file_save_01)
file.copy(path_out_01, path_in_01, overwrite = TRUE)

[1] TRUE

2. Metadatos de infraestructuras

In [21]:
data_to_save_02 <- tdata_02

In [22]:
archivo <- "_02"
proper <- "_pois_meta"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [ ]:
# file_save_02 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_02 <- paste0(oPath, file_save_02)
# write_csv(data_to_save_02, path_out_02)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [23]:
file_save_02 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_02 <- paste0(oPath, file_save_02)
write_csv(data_to_save_02, path_out_02)

cat('File saved as: ')
path_out_02

File saved as: 

[1] "Data/Output/CU_18_05_13_02_pois_meta.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [24]:
path_in_02 <- paste0(iPath, file_save_02)
file.copy(path_out_02, path_in_02, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

This working code needs the following conditions:

* For using the interactive selection of file, the {tcltk} package must be installed. It is not needed in production.
* The {readr}, {dplyr}, {tidyr} and {janitor} packages must be installed. 
* The data paths `Data/Input` and `Data/Output` must exist (relative to the notebook path)

### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* tidyr 1.3.0
* dplyr 1.0.10
* janitor 2.1.0
* readr 2.1.3

### Data structures

#### Objeto `tdata_01`

- Tenemos 224 filas, una por distrito, con los recuentos de cada uno de los tipos de pois (58)



In [25]:
glimpse(tdata_01)

Rows: 224
Columns: 60
$ cmun              <chr> "002", "003", "004", "005", "005", "005", "005", "00~
$ cdis              <chr> "01", "01", "01", "01", "02", "03", "04", "05", "01"~
$ bakery            <int> 2, 0, 0, 4, 5, 1, 2, 7, 4, 1, 1, 0, 2, 0, 0, 1, 0, 0~
$ bar               <int> 2, 0, 1, 22, 10, 2, 3, 11, 7, 1, 1, 2, 6, 0, 2, 17, ~
$ butcher           <int> 1, 0, 0, 2, 2, 0, 1, 2, 0, 0, 1, 0, 0, 0, 0, 2, 0, 0~
$ cafe              <int> 5, 1, 0, 12, 6, 1, 3, 19, 12, 3, 2, 0, 6, 0, 1, 4, 0~
$ car_dealership    <int> 1, 0, 0, 1, 1, 3, 0, 3, 5, 1, 0, 0, 0, 0, 0, 0, 0, 0~
$ clothes           <int> 1, 0, 0, 13, 1, 0, 1, 28, 0, 0, 0, 0, 0, 0, 1, 1, 0,~
$ convenience       <int> 2, 0, 0, 4, 2, 2, 2, 13, 8, 1, 1, 2, 6, 0, 0, 1, 0, ~
$ fast_food         <int> 1, 0, 0, 8, 1, 2, 5, 11, 10, 7, 4, 0, 6, 0, 2, 5, 0,~
$ food_court        <int> 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0~
$ furniture_shop    <int> 2, 0, 0, 3, 0, 0, 1, 8, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0~
$ gift_shop       

## Objeto `tdata_02`

- Tenemos 58 filas, una por tipo de poi, con sus metadatos



In [26]:
glimpse(tdata_02)

Rows: 58
Columns: 4
$ desc_var   <chr> "bakery", "bar", "butcher", "cafe", "car_dealership", "clot~
$ nombre_var <chr> "bakery", "bar", "butcher", "cafe", "car_dealership", "clot~
$ grupo      <chr> "comercio", "hosteleria", "comercio", "hosteleria", "comerc~
$ n          <int> 627, 2044, 289, 1577, 261, 1435, 1200, 1049, 24, 286, 129, ~


#### Observaciones generales sobre los datos

* No aplica

### Consideraciones para despliegue en piloto

* Utilizar los metadatos para agrupar variables si hiciera falta.

### Consideraciones para despliegue en producción

* No aplica

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han agrupado las infraesctructuras por distrito
- Se han calculado el número de infraestructuras por tipo en cada distrito

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben unir al resto de datos por distrito

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [ ]:
# incluir código